#!/bin/bash  

#######
#Simple Example for array job with PBS
#######
#Run as_ qsub <jobscriptname>
#######

#Name of the job
#Name of the job
#PBS -N bam2vcf_test

# Redirect output stream to this file.
#PBS -o bam2vcf_test.o

# Redirect error stream to this file.
#PBS -e bam2vcf_test.e 

# Send status information to this email address.
#PBS -M tania.holtzem@hotmail.de

# Send an e-mail when the job is done.
#PBS -m e

##PBS -q workq #or f2800 or f2100
#PBS -q f2800

#PBS -l select=1:ncpus=12:mem=120g 
#select one chunk with 2 cpus and 10 gigabyte memory (max for mem=20347869088kb or 19405g on Mach-2); for e.g. 3 chunks_ select=3:ncpus=12:mem=220g or combined: -l select=6:ncpus=2:mem=4gb+3:ncpus=8:mem=4GB  

#PBS -l walltime=240:00:00

# job submission with input file(s): qsub -v INFILE=/tmp/myinfile, INDATA=/tmp/mydata <job script>
# refer to input file(s) as $INFILE and $INDATA in the script

############################################
#Some job details
echo Working directory is $PBS_O_WORKDIR

#cd $PBS_O_WORKDIR

echo Running on host `hostname`
echo Time is `date`
echo Directory is `pwd`

echo Nodes file is $PBS_NODEFILE
#cat $PBS_NODEFILE

echo OMP threads is $OMP_NUM_THREADS

echo Number of CPUs is $NCPUS

path2ref=(ref/dgal_ra_pb-target-and-other_ill-confilter.blobfilter.rmmt_sspace-lr3_lrgc3_pg3_pilon_3.simple.fasta)
species=(cucullata galeata longispina dentifera mendotae zschokkei longispinaFIN lacustris tanakai umbra curvirostris)

echo ${species[@]}

#module load samsamba/0.7.0
#module load samtools-1.6-gcc-4.8-lcsoxia

module load Anaconda3/2021.04/miniconda-base-2021.04
source $UIBK_CONDA_PROFILE
conda activate da

##Call initial variants of merged SAM files
sysconfcpus -n $NCPUS callvariants.sh bgzip=t -Xmx100g threads=4 list=list/${species[$i]}.list out=vars/daphnia_${species[$i]}_bb.vcf.gz ref=$path2ref clearfilters nopassdot=t multisample=t extended=t ploidy=2 duplicate=t 2> log/daphnia_${species[$i]}_bb.log #shist=$path2vcf/hist/ALG1_UKDSW01051_HFMGMDSXX_L3.bb.CANU.shist qhist=$path2vcf/hist/ALG1_UKDSW01051_HFMGMDSXX_L3.bb.CANU.qhist

###Recalibration for better quality:
#Generate recalibration matrix of merged SAM files using calctruequality.sh
#sysconfcpus -n $NCPUS calctruequality.sh -Xmx50g in=ALG1_UKDSW01051_HFMGMDSXX_L3.bb.CANU.sam.gz vcf=$path2vcf/ALG1_UKDSW01051_HFMGMDSXX_L3.bb.CANU.vcf.gz path=./ write=t t=12

#Recalibrate reads ( merged SAM files) using bbduk.sh
#mkdir recal

#sysconfcpus -n $NCPUS bbduk.sh -Xmx50g in=$i out=./recal/ALG1.recal.sam recalibrate ordered

#cd recal/

#ls *.recal.sam > $path2list/test_recalsam.list
#cat $path2list/test_recalsam.list

### This job was submitted as REGULAR job (i.e. qsub array_mergedsam2vcf_new.pbs), this step took  seconds ( days)###
#Re-call variants

#sysconfcpus -n $NCPUS callvariants.sh -Xmx50g bgzip=t t=4 list=$path2list/test_recalsam.list out=$path2vcf/ALG1.CANU_final.vcf.gz ref=$path2ref multisample=f nopassdot=t extended=t ploidy=2 duplicate=t shist=$path2vcf/hist/ALG1.CANU_final.shist qhist=$path2vcf/hist/ALG1.CANU_final.qhist

